lib/portage/process.py

```diff
@@ -499,26 +503,37 @@ def _exec(binary, mycommand, opt_name, fd_pipes, env, gid, groups, uid, umask,
 			f.write('%d\n' % os.getpid())
 
 	# Unshare (while still uid==0)
-	if unshare_net or unshare_ipc:
+	if unshare_net or unshare_ipc or unshare_pid:
 		filename = find_library("c")
 		if filename is not None:
 			libc = LoadLibrary(filename)
 			if libc is not None:
+				# from /usr/include/bits/sched.h
 				CLONE_NEWIPC = 0x08000000
+				CLONE_NEWPID = 0x20000000
 				CLONE_NEWNET = 0x40000000
 
 				flags = 0
 				if unshare_net:
 					flags |= CLONE_NEWNET
 				if unshare_ipc:
 					flags |= CLONE_NEWIPC
+				if unshare_pid:
+					flags |= CLONE_NEWPID
 
 				try:
 					if libc.unshare(flags) != 0:
 						writemsg("Unable to unshare: %s\n" % (
 							errno.errorcode.get(ctypes.get_errno(), '?')),
 							noiselevel=-1)
 					else:
+						if unshare_pid:
+							# pid namespace requires us to fork() again
+							fork_ret = os.fork()
+							if fork_ret != 0:
+								pid, status = os.waitpid(fork_ret, 0)
+								assert pid == fork_ret
+								os._exit(status)
```

Comment:
What we really want is here is exactly what [dumb-init](https://github.com/Yelp/dumb-init) does, it calls waitpid on zombie child processes and also forwards signals that it receives to the child processes. For example, this would completely handle it:
```diff
--- a/lib/portage/process.py
+++ b/lib/portage/process.py
@@ -462,3 +467,9 @@ def _exec(binary, mycommand, opt_name, fd_pipes, env, gid, groups, uid, umask,
 	# Set up the command's argument list.
-	myargs = [opt_name]
+	if unshare_pid:
+		# pid namespace requires an init process
+		# reaps children and forwards signals
+		myargs = [opt_name, binary]
+		binary = '/usr/bin/dumb-init'
+	else:
+		myargs = [opt_name]
 	myargs.extend(mycommand[1:])
```
Here's some sample python code to handle the zombie child processes, but doesn't forward signals to child processes:
```diff
--- a/lib/portage/process.py
+++ b/lib/portage/process.py
@@ -519,6 +535,36 @@ def _exec(binary, mycommand, opt_name, fd_pipes, env, gid, groups, uid, umask,
 							errno.errorcode.get(ctypes.get_errno(), '?')),
 							noiselevel=-1)
 					else:
+						if unshare_pid:
+							# pid namespace requires an init-like process
+							# that calls waitpid on children and forwards
+							# signals like github.com/Yelp/dumb-init does.
+							# TODO: forward signals
+							fork_ret = os.fork()
+							if fork_ret != 0:
+								def handle_sigchld(_signum, _frame):
+									while True:
+										try:
+											pid, status = os.waitpid(-1, os.WNOHANG)
+										except OSError:
+											break
+										else:
+											if pid == fork_ret:
+												handle_sigchld.status = status
+
+								handle_sigchld.status = None
+								signal.signal(signal.SIGCHLD, handle_sigchld)
+								while handle_sigchld.status is None:
+									try:
+										pid, status = os.waitpid(fork_ret, 0)
+									except OSError as e:
+										if e.errno == errno.ESRCH:
+											break
+									else:
+										handle_sigchld.status = status
+
+								os._exit(1 if handle_sigchld.status is None
+									else handle_sigchld.status)
 						if unshare_net:
 							# 'up' the loopback
 							IFF_UP = 0x1
```